test/cluster/mzcompose.py

```diff
@@ -1077,3 +1078,116 @@ def fetch_reconciliation_metrics() -> Tuple[int, int]:
     # TODO(#17594): Flip these once the bug is fixed.
     assert reused == 0
     assert replaced == 4
+
+
+def workflow_test_mz_subscribes(c: Composition) -> None:
+    """
+    Test that in-progress subscribes are reflected in mz_subscribes.
+    """
+
+    c.down(destroy_volumes=True)
+
+    c.up("testdrive", persistent=True)
+    c.up("materialized", "clusterd1", "clusterd2")
+
+    c.sql(
+        """
+        CREATE CLUSTER cluster1 REPLICAS (r (
+                STORAGECTL ADDRESSES ['clusterd1:2100'],
+                STORAGE ADDRESSES ['clusterd1:2103'],
+                COMPUTECTL ADDRESSES ['clusterd1:2101'],
+                COMPUTE ADDRESSES ['clusterd1:2102'],
+                WORKERS 1
+        ));
+        CREATE CLUSTER cluster2 REPLICAS (r (
+                STORAGECTL ADDRESSES ['clusterd2:2100'],
+                STORAGE ADDRESSES ['clusterd2:2103'],
+                COMPUTECTL ADDRESSES ['clusterd2:2101'],
+                COMPUTE ADDRESSES ['clusterd2:2102'],
+                WORKERS 1
+        ));
+        CREATE TABLE t (a int);
+        """
+    )
+
+    def check_mz_subscribes(expected: Tuple) -> None:
+        """
+        Check that the expected subscribes exist in mz_subscribes.
+        We identify subscribes by user and cluster only. We explicitly don't
+        check the `GlobalId`, as how that is allocated is an implementation
+        detail and might change in the future.
+        """
+        output = c.sql_query(
+            f"""
+            SELECT s.user, c.name
+            FROM mz_internal.mz_subscribes s
+              JOIN mz_clusters c ON (c.id = s.cluster_id)
+            ORDER BY s.started_at
+            """
+        )
+        assert output == expected, f"expected: {expected}, got: {output}"
+
+    def run_subscribe(id_: int, on_cluster: str) -> None:
+        """
+        Run a subscribe with on the given cluster.
+        The subscribe tries to fetch one record with a value of `id_` from
+        table `t`, which means it can be stopped by INSERTing that value.
+        """
+        c.sql(
+            f"""
+            SET cluster = {on_cluster};
+            BEGIN;
+            DECLARE c CURSOR FOR SUBSCRIBE (SELECT * FROM t WHERE a = {id_});
+            FETCH 1 c WITH (TIMEOUT = '10s');
```

Comment:
Why is a timeout needed here?

Reply:
I noticed that if an assert fails in the main thread and the INSERT that makes the FETCH fail never happens, the testdrive job just hangs. The timeout is to make it stop hanging after 10s.

Reply:
I think either testdrive or rust tests handle this and the sleep issue more gracefully. In that they retry for a certain amount of time and then timeout eventually. If possible I think this test would be better in one of those two format.s

Reply:
I don't think one can use testdrive to create multiple concurrent sessions like we need for having live subscribes while also checking the contents of `mz_subscribes`. By rust tests do you mean the ones in `src/environmentd/tests`?

Reply:
Yeah, in `src/environmentd/tests`. If it's too difficult though, I can live with these tests.

Reply:
I've implemented the test now based on cursors, rather than threads, which has the nice side-effect that no timeouts are needed anymore. I think that addresses your concerns!